### PR TITLE
cloud/aws: allow users to provide role_arn in configuration

### PR DESCRIPTION
**Pull request by icy**
### What does this PR do?

Allow users to provide `role_arn` in configuration for `aws` cloud provider.

### Previous Behavior
Remove this section if not relevant

### New Behavior
Remove this section if not relevant

### Tests written?

TODO

### Commits signed with GPG?

Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/cloud/clouds/ec2.py [hunk 1]
```diff
@@ -16,6 +16,9 @@
       # EC2 metadata set both id and key to 'use-instance-role-credentials'
       id: GKTADJGHEIQSXMKKRBJ08H
       key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
+      # If 'role_arn' is specified the above credentials are used to
+      # to assume to the role.
+      role_arn: None
```
Comment (gtmanfred) [inline salt/cloud/clouds/ec2.py]: This needs to be documented in the ec2 cloud getting started page.

**Review by gtmanfred — CHANGES_REQUESTED** [review]

**Comment by icy** [comment]
Question 1: Why doesn't `Salt` get credentials from local files `~/.aws/credentials` instead? Use `use-instance-role-credentials` is also good however the `aws` way has a few more convenience. It seems `salt` tried to avoid boto style

Question 2: How to write new tests file for new feature?

**Comment by gtmanfred** [comment]
1) I don't know, that is the way it has always been.  I don't have a reason it couldn't be added to pull from ~/.aws/credentials is nothing is configured.

2) Here is the documentation for writing tests https://docs.saltstack.com/en/latest/topics/tutorials/writing_tests.html

**Comment by icy** [comment]
Thanks a lot @gtmanfred . I will take a look.

**Comment by icy** [comment]
Hi again,

do you have any idea to set up the tests correctly? I update the `key` and `id` fields in `ec2.conf` but I still got errors as below

```
# EXPENSIVE_TESTS=True python tests/runtests.py  -n integration.cloud.providers.test_ec2
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Python Version: 2.7.12 (default, Dec 4 2017, 14:50:18) [GCC 5.4.0 20160609]
 * Transplanting configuration files to '/tmp/salt-tests-tmpdir/config'
 * Current Directory: /root/salt
 * Test suite is running under PID 15716
 * Logging tests on /tmp/salt-runtests.log
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Max open files settings is too low (soft: 1024, hard: 1048576) for running the tests
 * Trying to raise the limits to soft: 3072, hard: 1048576
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Setting up Salt daemons to execute tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Starting salt-master ... STARTED!                                                                                                                                                                                                                                                      
 * Starting salt-minion ... STARTED!                                                                                                                                                                                                                                                      
 * Starting sub salt-minion ... STARTED!                                                                                                                                                                                                                                                  
 * Starting syndic salt-master ... STARTED!                                                                                                                                                                                                                                               
 * Starting salt-syndic ... STARTED!                                                                                                                                                                                                                                                      
==========================================================================================================================================================================================================================================================================================
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Starting integration.cloud.providers.test_ec2 Tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Traceback (most recent call last):
  File "/tmp/salt-tests-tmpdir/scripts/salt-cloud", line 2, in <module>
    from salt.scripts import salt_cloud
ImportError: No module named salt.scripts
sTraceback (most recent call last):
  File "/tmp/salt-tests-tmpdir/scripts/salt-cloud", line 2, in <module>
    from salt.scripts import salt_cloud
ImportError: No module named salt.scripts
sTraceback (most recent call last):
  File "/tmp/salt-tests-tmpdir/scripts/salt-cloud", line 2, in <module>
    from salt.scripts import salt_cloud
ImportError: No module named salt.scripts
ssTraceback (most recent call last):
  File "/tmp/salt-tests-tmpdir/scripts/salt-cloud", line 2, in <module>
    from salt.scripts import salt_cloud
ImportError: No module named salt.scripts
ss
----------------------------------------------------------------------
Ran 6 tests in 0.836s

OK (skipped=6)

=================================================================================================================================  Overall Tests Report  =================================================================================================================================
*** integration.cloud.providers.test_ec2 Tests  ******************************************************************************************************************************************************************************************************************************************
 --------  Skipped Tests  ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
   -> integration.cloud.providers.test_ec2.EC2Test.test_instance          ->  Configuration file for ec2 was not found. Check ec2.conf files in tests/integration/files/conf/cloud.*.d/ to run these tests.
   -> integration.cloud.providers.test_ec2.EC2Test.test_instance_rename   ->  Configuration file for ec2 was not found. Check ec2.conf files in tests/integration/files/conf/cloud.*.d/ to run these tests.
   -> integration.cloud.providers.test_ec2.EC2Test.test_win2012r2_psexec  ->  Configuration file for ec2 was not found. Check ec2.conf files in tests/integration/files/conf/cloud.*.d/ to run these tests.
   -> integration.cloud.providers.test_ec2.EC2Test.test_win2012r2_winrm   ->  Skip when winrm dependencies are missing
   -> integration.cloud.providers.test_ec2.EC2Test.test_win2016_psexec    ->  Configuration file for ec2 was not found. Check ec2.conf files in tests/integration/files/conf/cloud.*.d/ to run these tests.
   -> integration.cloud.providers.test_ec2.EC2Test.test_win2016_winrm     ->  Skip when winrm dependencies are missing
 -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
==========================================================================================================================================================================================================================================================================================
OK (total=6, skipped=6, passed=0, failures=0, errors=0) 
=================================================================================================================================  Overall Tests Report  =================================================================================================================================
```

Thanks alot

**Comment by gtmanfred** [comment]
I would highly recommend writing unit tests for this, instead of integration tests.

For setting up the cloud integration tests, we use this state.

https://github.com/saltstack/salt-jenkins/blob/master/cloud-test-configs/ec2.sls

It sets up the config for the ec2 cloud driver.

But yeah, if you write unit tests, that would be way better, cause they run more frequently, and don't require access to the cloud accounts.

**Comment by gtmanfred** [comment]
There are also some lint failures.

https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/22530/console

**Review by gtmanfred — APPROVED** [review]

**Comment by icy** [comment]
Thanks for your review and approval @gtmanfred . I will try to add some tests.

**Pull request merged**